File: FirstMeeting/StarterCode/edmonds_karp.java
import java.util.*;
import java.lang.*;
import java.io.*;
import java.util.LinkedList;

class EdmondsKarp {

    /* Returns true if there is a path from source 's' to sink
      't' in residual graph. Also fills path[] to store the
      path */
    boolean breathFirstSearch(int residual_graph[][], int s, int t, int path[]) {
        // Create a visited array and mark all vertices as not
        int numOfVertices = residual_graph.length;
        boolean visited[] = new boolean[numOfVertices];
        for(int i = 0; i < numOfVertices; ++i)
            visited[i]=false;

        // Create a queue, enqueue source vertex and mark
        // source vertex as visited
        LinkedList<Integer> queue = new LinkedList<Integer>();
        queue.add(s);
        visited[s] = true;
        path[s]=-1;

        // Standard BFS Loop
        while (queue.size()!=0) {
            int u = queue.poll();

            for (int v = 0; v < numOfVertices; v++) {
                if (visited[v] == false && residual_graph[u][v] > 0) {
                    queue.add(v);
                    path[v] = u;
                    visited[v] = true;
                }
            }
        }

        // If we reached sink in BFS starting from source, then
        // return true, else false
        return (visited[t] == true);
    }

    /* ********************* ignore the code above this line ******************** */


    void updateResidualGraph(int residual_graph[][], int source, int sink, int path[], int path_flow) {
        for (int v = sink; v != source; v = path[v]) {
            int u = path[v];
            /* each iteration gets you a pair of vertices u and v between an edge in the path
             * so residual_graphp[u][v] is the residual capacity of an edge in the path
             * */

            /* **** your code here **** */
        }
    }

    int findMinimalResidualCapacity(int residual_graph[][], int source, int sink, int path[]) {
        int path_flow = Integer.MAX_VALUE;
        for (int v = sink; v != source; v = path[v]) {
            int u = path[v];
            /* each iteration gets you a pair of vertices u and v between an edge in the path
             * so residual_graphp[u][v] is the residual capacity of an edge in the path
             * */

            /* **** your code here **** */

        }
        return path_flow;
    }

    /* source and sink are the index of source and sink node */
    int edmondsKarp(int input_graph[][], int source, int sink) {
        int numOfVertices = input_graph.length;

        /* create a residual_graph by copying the capacity graph */
        int RG[][] = new int[numOfVertices][numOfVertices];
        for (int u = 0; u < numOfVertices; u++)
            for (int v = 0; v < numOfVertices; v++)
            	RG[u][v] = input_graph[u][v];

        /* the vector to save the new path computed by BFS */
        int path[] = new int[numOfVertices];
        int max_flow = 0;

        /* ****your code here**** */

        return max_flow;
    }


    public static void main (String[] args) {
        int graph[][] =new int[][] { {0, 16, 13, 0, 0, 0},
                                     {0, 0, 10, 12, 0, 0},
                                     {0, 4, 0, 0, 14, 0},
                                     {0, 0, 9, 0, 0, 20},
                                     {0, 0, 0, 7, 0, 4},
                                     {0, 0, 0, 0, 0, 0}
                                   };

        EdmondsKarp m = new EdmondsKarp();

        System.out.println("The maximum possible flow is " +
                           m.edmondsKarp(graph, 0, 5));
    }
}

File: FirstMeeting/StarterCode/edmonds_karp.cpp
#include <iostream>
#include <limits.h>
#include <string.h>
#include <queue>
#include <vector>
#include <bitset>
#include <algorithm>
using namespace std;
 
/* returns false if no path can be found from source to sink */ 
/* it updates vector<int>& path therefore the vector variable passed in the function 
 * have a new path after each iteration */ 
bool BreathFirstSearch(const vector<vector<int> > &residual_graph, 
        const int source, const int sink, vector<int> & path) {

    const int num_vertices = residual_graph.size();
    vector<bool> visited(num_vertices, false);

    queue <int> q;
    q.push(source);
    visited[source] = true;

    fill(path.begin(), path.end(), 0);
    path[source] = -1;
 
    while (!q.empty()) {
        int u = q.front();
        q.pop();

        for (int v = 0; v < num_vertices; v++) {
            if (visited[v] == false && residual_graph[u][v] > 0) {
                q.push(v);
                path[v] = u;
                visited[v] = true;
            }
        }
    }
    
    return (visited[sink] == true);
}

int FindMinimalResidualCapacity(const vector<vector<int> >& residual_graph, 
        const int source, const int sink, const vector<int>& path) {

    int path_flow = INT_MAX,
        pathlen = path.size();

    for (int v = sink; v != source; v = path[v]) { 
        int u = path[v];

        /* each iteration gets you a pair of vertices u and v between an edge in the path 
         * so residual_graphp[u][v] is the residual capacity of an edge in the path 
         * */

        /* **** your code here **** */

    }

    return path_flow;
}

void UpdateResidualGraph(vector<vector<int> >& residual_graph, 
        const int source, const int sink, const vector<int>& path, int path_flow) {

    int pathlen = path.size();
    for (int v = sink; v != source; v = path[v]) { 
        int u = path[v];

        /* each iteration gets you a pair of vertices u and v between an edge in the path 
         * so residual_graphp[u][v] is the residual capacity of an edge in the path 
         * */

        /* **** your code here **** */
    }
}

/* source and sink are the index of source and sink node */
int Edmond_Karp(const std::vector<vector<int> > & residual_graph, 
        const int source, const int sink) {

    int max_flow = 0;

    /* copy the input to a new 2d vector */
    std::vector<vector<int> > RG(residual_graph);

    /* initalize the path vector to be zero 
     * you will pass this vector to bfs */
    std::vector<int> path(residual_graph.size(), 0);

    /* **** your code here **** */

    return max_flow;
}

int main(void) {
    int num_vertices, source, sink;

    cin >> num_vertices;
    cin >> source >> sink;

    /* initalize 2 d vector with all zeros */
    vector<vector<int> > graph(num_vertices, vector<int>(num_vertices, 0));

    for (int i = 0; i < num_vertices; i++) {
        for (int j = 0; j < num_vertices; j++) {
            cin >> graph[i][j];
        }
    }
     
    std::cout << Edmond_Karp(graph, source, sink) << std::endl;
}
